main-settings/src/main/scala/sbt/Scope.scala

```diff
@@ -25,15 +26,33 @@ final case class Scope(
   def rescope(config: ConfigKey): Scope = copy(config = Select(config))
   def rescope(task: AttributeKey[?]): Scope = copy(task = Select(task))
 
+  def copy(
+      project: ScopeAxis[Reference] = this.project,
+      config: ScopeAxis[ConfigKey] = this.config,
+      task: ScopeAxis[AttributeKey[?]] = this.task,
+      extra: ScopeAxis[AttributeMap] = this.extra
+  ): Scope = Scope(project, config, task, extra)
+
   override def toString: String = this match
     case Scope(Zero, Zero, Zero, Zero) => "Global"
     case Scope(_, _, _, This)          => s"$project / $config / $task"
     case _                             => s"Scope($project, $config, $task, $extra)"
 end Scope
 
 object Scope:
-  val ThisScope: Scope = Scope(This, This, This, This)
-  val Global: Scope = Scope(Zero, Zero, Zero, Zero)
+  private val uniquenessCache = TrieMap.empty[Scope, Scope]
```

Comment:
We have a few places in our code base that's using Caffeine (https://github.com/ben-manes/caffeine). If it's not load-bearing to enforce uniqueness, maybe it's safer to use something like caffeine, which would use weak reference, which allows GC to pick clean them up if necessary.

Reply:
Interesting! I am going to try it for `Scope`, but it also gives me some hope of reducing the number of long-living `ScopedKey`, which I'll try in another PR.

Reply:
It seems I cannot use caffeine with weak reference as a uniqueness cache, because it uses `eq` instead of `equals`, which defeats the purpose of a uniqueness cache. As an alternative I can create my own implementation of a uniqueness cache with weak reference, which would be optimized in terms of heap and CPU usage, but it is probably not worth it for `Scope`. I'll try that on `ScopedKey` in a different PR.

The uniqueness cache on `Scope` is already safe because the number of unique `Scope` is low compared to the size of the graph. It cannot explode.

Reply:
Perhaps [Interner](https://github.com/ben-manes/caffeine/wiki/Interner) is what you were looking for?

Reply:
Yes indeed! Thanks @ben-manes